### ISSUE TITLE
GitVersion 6.0.0 is not supported due to breaking changes

### ISSUE THREAD

**Issue by johlju** [gaelcolas/Sampler#477]
### Problem description

There are breaking changes in GitVersion that fails builds in projects using Sampler. Workaround is to pin an older version.

Notable breaking changes are configuration properties that has changed name, and output properties (`NuGetVersionV2`) that no longer exist.

Configuration documentation: https://gitversion.net/docs/reference/configuration
Output variables: https://gitversion.net/docs/reference/variables

### Verbose logs

```text
You can invoke the tool using the following command: dotnet-gitversion
Tool 'gitversion.tool' (version '6.0.0') was successfully installed.
ConvertFrom-Json : Conversion from JSON failed with error: Error parsing Infinity value. Path '', line 1, position 1.
At D:\a\_temp\87ce70e7-6735-4619-b9cf-c7cfe9c66d84.ps1:5 char:41
+ $gitVersionObject = dotnet-gitversion | ConvertFrom-Json
+                                         ~~~~~~~~~~~~~~~~
+ CategoryInfo          : NotSpecified: (:) [ConvertFrom-Json], ArgumentException
+ FullyQualifiedErrorId : System.ArgumentException,Microsoft.PowerShell.Commands.ConvertFromJsonCommand
##[error]PowerShell exited with code '1'.
```

### How to reproduce

Build a project using GitVersion v6.0.0

### Expected behavior

Sampler supports v6.0.0

### Current behavior

Fails if GitVersion 6.0.0 is installeed.

### Suggested solution

Update all tasks and configuration (and templates) to use another property than `NuGetVersionV2`.

Did a quick test in another project to see what was needed when I realized that we have hardcoded `NuGetVersionV2` in the tasks. This PR shows the needed config changes: https://github.com/viscalyx/PesterConverter/pull/28

### Operating system the target node is running

```text
n/a
```

### PowerShell version and build the target node is running

```text
n/a
```

### Module version used

```text
n/a
```

**Comment by johlju** [comment]
Workaround is to install 5.12.0, e.g.:

```
dotnet tool install --global GitVersion.Tool --version 5.12.0
```

**Comment by dan-hughes** [comment]
Can use `dotnet tool install --global GitVersion.Tool --version 5.*` in case any updates to v5 are released.

**Comment by johlju** [comment]
As mentioned by @gaelcolas offline is that we need to pad the preview string number with zeros, e.g. 'preview0001'. This so sorting in certain NuGet repositories work as expected, it is an OData issue with NuGet. To prevent that `-preview9` is assumed to be higher than `-preview12`.

Also, any dot (`.`) in the prerelease string must be removed as it is not supported in the module manifest (I might be wrong here).

This is an example from GitVersion's docs (link above):
```json
{
    "Major": 3,
    "Minor": 22,
    "Patch": 11,
    "PreReleaseTag": "beta.99",
    "PreReleaseTagWithDash": "-beta.99",
    "PreReleaseLabel": "beta",
    "PreReleaseLabelWithDash": "-beta",
    "PreReleaseNumber": 99,
    "WeightedPreReleaseNumber": 1099,
    "BuildMetaData": 88,
    "FullBuildMetaData": "99.Branch.release/3.22.11.Sha.28c853159a46b5a87e6cc9c4f6e940c59d6bc68a",
    "MajorMinorPatch": "3.22.11",
    "SemVer": "3.22.11-beta.99",
    "AssemblySemVer": "3.22.11.0",
    "AssemblySemFileVer": "3.22.11.0",
    "InformationalVersion": "3.22.11-beta.99+88.Branch.release/3.022.011.Sha.28c853159a46b5a87e6cc9c4f6e940c59d6bc68a",
    "FullSemVer": "3.22.11-beta.99+88",
    "BranchName": "release/3.022.011",
    "EscapedBranchName": "release-3.022.011",
    "Sha": "28c853159a46b5a87e6cc9c4f6e940c59d6bc68a",
    "ShortSha": "28c8531",
    "VersionSourceSha": "28c853159a46b5a87e6cc9c4f6e940c59d6bc68a",
    "CommitsSinceVersionSource": 7,
    "CommitDate": "2021-12-31",
    "UncommittedChanges": 0
}
```

My thoughts:

- Build pipeline version title
  - The build pipeline should preferably use `FullSemVer` so that the pipeline title is unique for each build. _Not tested that it works._
- Module version
  - The module version should use the properties should concatenate `MajorMinorPatch` + `PreReleaseTag`.
  - The PreReleaseTag should be padded with 4 zeros and have the `.` removed, e.g.:
    ```powershell
    $prereleaseString = 'beta.99' # result from GitVersion's variable PreReleaseTag
    $match = [System.Text.RegularExpressions.Regex]::Match($prereleaseString, '(\d+)$')
    $numericValue = $match.Groups[1].Value.PadLeft(4, '0')
    $nonNumericPart = $prereleaseString.Substring(0, $match.Groups[1].Index) -replace '\.'
    $paddedPrereleaseString = "$nonNumericPart$numericValue"
    $paddedPrereleaseString
    ```

**Comment by johlju** [comment]
Maybe we should introduce a a command that has several purposes (maybe we already have one)

- Call gitversion (build script supports both `gitversion` and `donet-gitversion`)
- Return the ModuleVersion according to above
- Return the pipeline job title (semantic version + build number)
- Set the task variables for the Azure Pipelines pipeline (see azure-pipelines.yml)

The command can return different values depending on passed parameters.

**Comment by pshamus** [comment]
Another option I'm using is to add a build task to install GitVersion if needed:

```powershell
param (
    # Build Configuration object
    [Parameter()]
    [System.Collections.Hashtable]
    $BuildInfo = (property BuildInfo @{ })
)

# Synopsis: Installs GitVersion as a .NET Global Tool if needed
task Install_GitVersion -if ($BuildInfo.GitVersion.Version) {
    $dotNetGitVersionAvailable = Get-Command -Name 'dotnet-gitversion' -ErrorAction 'SilentlyContinue'

    # If neither dotnet-gitversion nor gitversion are available, install dotnet-gitversion.
    if (-not $dotNetGitVersionAvailable) {
        Write-Build -Color Green -Text "Installing GitVersion $($BuildInfo.GitVersion.Version) as a .NET Global Tool"
        $null = dotnet tool install --global GitVersion.Tool --version $BuildInfo.GitVersion.Version
    }
}

```

Once Sampler is updated to support both versions when determining the module version, you can selectively update a given project to use 6.x, which might also include updatin GitVersion.yml as the schema has changed too.

**Comment by johlju** [comment]
> Another option I'm using is to add a build task to install GitVersion if needed

Downside to this is that we shouldn't change the users machine which will happen if the user runs the task locally. Currently Sampler expects correct version to be installed on the users machine. Is it possible to save it just for the project without installing it, like portable so it can be easily removed?

**Comment by johlju** [comment]
For macOS and `brew` users:

It is possible to run `$ModuleVersion = '0.0.1'` locally before build to skip GitVersion requirement if it is not installed and there is no need to build using correct version. 🤔 

I currently need to do that on my macOS which updated to v6, where it is not possible to revert to 5.x version using `brew` (I can't find a way for GitVersion at least).

**Comment by pshamus** [comment]
If you install the tool in a custom path (which I am currently doing), you can invoke from there. So maybe the tool is installed in `output`, that way it's not installed on the user's machine outside of each repository. You'd need to update the build scripts to invoke `dotnet-gitversion` from the custom path. But it's a nice way to not rely on the pipeline to determine what version to use so you can reproduce the behavior locally.

**Comment by johlju** [comment]
> So maybe the tool is installed in output, that way it's not installed on the user's machine outside of each repository.

I wonder what is required to get `dotnet` working on a stock *nix, macOS and Windows. 🤔 Maybe installing PowerShell is enough for `dotnet` tool to work? I think your suggest could work, only adding a task if the user says in the template that GtVersion should be used to handle versioning. 🤔

**Comment by gaelcolas** [comment]
Waiting on #529 from @raandree

**Comment by Princetimber** [comment]
Still reproduces on **Sampler 0.120.1** (current latest) with **GitVersion 6.8.2**, so recording a fresh confirmation plus two things that may save the next person some time.

### Hardcoded `NuGetVersionV2` sites in 0.120.1

```
Sampler.psm1:1123                                    # Get-SamplerBuildVersion (lines 1091-1155)
Templates/Git/GitVersion.yml.template:7              # assembly-informational-format
Templates/Sampler/azure-pipelines.yml.template:68
Templates/Sampler/azure-pipelines_dsccommunity.yml.template:68
```

`Get-SamplerBuildVersion` is the one that bites locally — it is reached whenever GitVersion is on `PATH` and `ModuleVersion` was not supplied via the environment, which is every developer running `./build.ps1` directly. In CI the version usually arrives via `$env:ModuleVersion`, so a project can look migrated while local builds are silently broken.

### `SemVer` is not a safe substitute, and the failure is deferred to publish time

Following up on the note about `'.'` not being valid in a module manifest — the downstream behaviour is worse than a hard failure. Feeding v6's `SemVer` through the build does **not** error:

| `$env:ModuleVersion` | `Split-ModuleVersion` → prerelease | Built manifest | Valid? |
|---|---|---|---|
| `0.2.0-preview.1` (v6 `SemVer`) | `preview.1` | `Prerelease = 'preview_1'` | ❌ underscore |
| `0.2.0-preview0001` | `preview0001` | `Prerelease = 'preview0001'` | ✅ |

The `.` is rewritten to `_`, the build reports success, and **`Test-ModuleManifest` accepts `preview_1`**. A PowerShell prerelease must be alphanumeric, so the first genuine failure is at `Publish-Module` / PSGallery — after a release build has otherwise gone green.

### A composition that satisfies both constraints

Since the requirement is a prerelease with no `.` and a zero-padded counter (for the NuGet OData sort ordering mentioned above), composing from `MajorMinorPatch` + `PreReleaseLabel` + `PreReleaseNumber` reproduces the historical `NuGetVersionV2` shape without needing the removed variable:

```powershell
if ([string]::IsNullOrWhiteSpace($preReleaseLabel)) {
    # Tagged release: bare version, no prerelease segment.
    $moduleVersion = $majorMinorPatch
}
else {
    # Prerelease strings must be alphanumeric; the counter is padded so
    # -preview0012 sorts above -preview0009.
    $label  = $preReleaseLabel -replace '[^A-Za-z0-9]', ''
    $number = if ([string]::IsNullOrWhiteSpace($preReleaseNumber)) { 0 } else { [int] $preReleaseNumber }
    $moduleVersion = '{0}-{1}{2:d4}' -f $majorMinorPatch, $label, $number
}
```

Verified against GitVersion 6.8.2 on a Sampler-based repo, all four branch classes producing manifest-valid versions:

| Context | Composed | Prerelease valid |
|---|---|---|
| `main` (`label: preview`) | `0.2.0-preview0047` | ✅ |
| tag `v0.3.0` | `0.3.0` | ✅ (no prerelease) |
| `pull-request` (`label: PR`) | `0.2.0-PR0047` | ✅ |
| `feature/thing` (`label: '{BranchName}'`) | `0.2.0-thing0001` | ✅ |

Sanitising the label also covers branch names that would otherwise leak `.` or `-` into the prerelease via `{BranchName}`.

### Config migration notes

For the `GitVersion.yml` side, the changes needed were: branch key `master` → `main`; per-branch `tag:` → `label:`; `useBranchName` → `label: '{BranchName}'`, which additionally requires the branch `regex` to expose a named `(?<BranchName>.+)` capture group or the label resolves empty; and `assembly-informational-format` can no longer reference `{NuGetVersionV2}` — v6 fails with `'NuGetVersionV2' is not a property or field on type 'SemanticVersionFormatValues'`. Top-level `mode:` is unchanged.

One behavioural difference worth flagging for anyone migrating: in a repository with no version tags, v6 treats `next-version` as authoritative and did not apply loose commit-message bump rules that v5 honoured (`minor-version-bump-message: '(adds?|features?|minor)\b'`), so a base version that had crept to `0.2.0` under v5 came back as `0.0.1` under v6. Pinning `next-version` restores it, but it changes the versioning model from commit-message-driven to tag-driven.

For now we are staying on `gittools/actions/gitversion@v3` with `versionSpec: '5.x'`, per the workaround above.
